pkg/reconciler/certificate/certificate.go

```diff
@@ -110,15 +112,17 @@ func (c *Reconciler) reconcile(ctx context.Context, knCert *v1alpha1.Certificate
 	knCert.Status.InitializeConditions()
 
 	logger.Infof("Reconciling Cert-Manager certificate for Knative cert %s/%s.", knCert.Namespace, knCert.Name)
+	knCert.Status.ObservedGeneration = knCert.Generation
+
 	cmConfig := config.FromContext(ctx).CertManager
 	cmCert := resources.MakeCertManagerCertificate(cmConfig, knCert)
 	cmCert, err := c.reconcileCMCertificate(ctx, knCert, cmCert)
 	if err != nil {
+		knCert.Status.MarkNotReady(notReconciledReason, notReconciledMessage)
```

Comment:
We might want to include `err` in the message, WDYT?

I'm also worried that this isn't robust to future changes -- if we add another path to `reconcile` that can return an error, we'll need to duplicate this `MarkNotReady` call. I'd move this up into `Reconcile` and check it immediately after we call `reconcile`.

Reply:
I am not sure if we want to include error message into Status because sometimes error message could be very verbose. Having it in the status may cause the Object exceeds k8s size limit.

Reply:
Good point... it would still be nice to have *something*, maybe we could truncate it?

Reply:
It sounds good to me to put the truncated error message into the status message. We probably also want to add some description before the truncated error message just in case that the truncated error message is not descriptive enough. (e.g. have status message like: fmt.Sprintf("Cert-Manager certificate reconciliation failed: %s", truncatedErrorMessage)).

Also do we want to consider as a general pattern to follow for `MarkFailed` or `MarkNotReady` when they are caused by errors?

Reply:
How much should we truncate the err message?

Reply:
Thought about this again. I think we probably don't want to put the error message into the status even after it is truncated. (Sorry for the back and forth)
Consider a scenario that the error message (or the truncated error message) will be changed every time when reconciling the same child resource object. (This could happen if the child resource reconciler add the error timestamp into the error message).
For this scenario, every time when reconciling the same child object (CM certificate in this case), the status of parent object (Knative Certificate in this case) will be changed, which will trigger the parent object [enqueuing](https://github.com/knative/serving/blob/105bc960934f186014f2193e4814c5a7cc700e06/pkg/reconciler/certificate/controller.go#L62) immediately . This process keeps repeating and won't be ended. But the ideal behavior for handing error should be [enqueuing the parent object with rate limited](https://github.com/knative/serving/blob/105bc960934f186014f2193e4814c5a7cc700e06/pkg/reconciler/certificate/controller.go#L62). The idea error-handle process will be broken in this scenario.

Reply:
I believe the hard limit is 1MB for an etcd entry? Let's start with something conservative like 255. If our error messages are too long, we can revisit them?

It would be nice to also indicate the message was truncated, so something like:
```go
const limit = 255

// ...

msg := fmt.Sprintf("Certificate reconciliation failed: %v", err)
if len(msg) > limit {
	msg = msg[:limit-3] + "..."
}
```

which would add "..." to the end of the message if we have to truncate it.

Reply:
> Consider a scenario that the error message (or the truncated error message) will be changed every time when reconciling the same child resource object. (This could happen if the child resource reconciler add the error timestamp into the error message).

Hmm... yeah this wouldn't be great. A timestamped error at a leaf object would cause a lot of churn in the condition tree.

> But the ideal behavior for handing error should be enqueuing the parent object with rate limited. The idea error-handle process will be broken in this scenario.

When a reconciler returns an error, that object's re-enqueue is actually rate limited:
https://github.com/knative/serving/blob/105bc960934f186014f2193e4814c5a7cc700e06/vendor/knative.dev/pkg/controller/controller.go#L344-L354

We use the default rate limiter:
https://github.com/knative/serving/blob/105bc960934f186014f2193e4814c5a7cc700e06/vendor/knative.dev/pkg/controller/controller.go#L146

Which has per-item exponential backoff:
https://github.com/knative/serving/blob/105bc960934f186014f2193e4814c5a7cc700e06/vendor/k8s.io/client-go/util/workqueue/default_rate_limiters.go#L37-L45

With a maximum of 1000 seconds, or about 15 minutes -- are we okay with statuses churning every 15 minutes?

Reply:
Nevermind, @ZhiminXiang is right. When we update the certificate status, it gets re-enqueued (not rate limited). Bummer.

Reply:
Alright, we can fix this: https://github.com/knative/serving/blob/d7426fa29fdfa007b532c22055691c332a6f5e36/pkg/reconciler/certificate/controller.go#L62

HandleAll currently ignores the old object in its UpdateFunc:
https://github.com/knative/serving/blob/d7426fa29fdfa007b532c22055691c332a6f5e36/vendor/knative.dev/pkg/controller/controller.go#L79

If we have a different UpdateFunc that only passes the new object if old.Spec != new.Spec, I think that should work... we care about status updates for child resources, but not for the certificate itself.

Reply:
@ZhiminXiang @mattmoor here's a quick and dirty prototype of what I was thinking: https://github.com/knative/serving/compare/master...jonjohnsonjr:controller-prototype

We probably also want to check for metadata changes as well...

Matt had an idea to do this in the webhook somehow?

Reply:
Thanks for the deep diving @jonjohnsonjr 
1. Generally I think we should reconcile an object when its status was changed because the status change of the object could be caused by the status change of its child resources, which means we need to enqueue the object to do "something".
2. If we need some fields in the status for the informative purpose, we could exclude these fields when [comparing](https://github.com/knative/serving/compare/master...jonjohnsonjr:controller-prototype#diff-c707780e22f7d988b1ae28b884067818R88) the new and old objects.
Thoughts?

Reply:
> Generally I think we should reconcile an object when its status was changed because the status change of the object could be caused by the status change of its child resources, which means we need to enqueue the object to do "something".

But the Foo reconciler is the thing that is updating the FooStatus, so we don't need to trigger the Foo reconciler on FooStatus updates. The Foo reconciler gets triggered by children updates via children informers, which is what caused the Foo reconcile to update the FooStatus in the first place :) When child statuses change, we should definitely re-reconcile the parent, but when our own status changes, we are triggering ourselves, which should always be a no-op unless we've made a huge mistake.

Reply:
Alright, so I think it's possible to make this work, but let's drop `err` from the message for now until we've landed some changes to make it feasible. Sorry for the back and forth :)

Reply:
Sounds good. Should the enqueuing logic remain the same for now?

Reply:
Yeah, the status update should trigger one extra reconcile loop (as it does already, today), but since the error message will remain constant, we won't get stuck in a reconcile loop.